Remove SharedVariableMap from server session

rogersteblerbsi opened the pull request:
As part of the move toward a stateless server session, the m_sharedVariableMap has been removed from the server session.

The SharedVariableMap has been renamed to VariableMap and is now used exclusively on the client side.

Instead of synchronizing variables from server to client, the client session initializes its VariableMap using the ISessionService.

To support updates after initialization, the server can send a client notification. A corresponding handler on the client side should process the notification and update the relevant variables on the client session.

423389

rogersteblerbsi committed:
Remove SharedVariableMap from server session

As part of the move toward a stateless server session, the m_sharedVariableMap has been removed from the server session.

The SharedVariableMap is now used exclusively on the client side.

Instead of synchronizing variables from server to client, the client session initializes its VariableMap using the ISessionService.

To update variables after initialization, the server can send a SharedContextChangedNotification to the client.

423389

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
    */
-  protected void initializeSharedVariables() {
-    BEANS.get(IPingService.class).ping("");
+  protected void loadAndInitializeVariables() {
```
Who is calling this method? Shouldnt this be called in #execLoadSession()
Unit Test?

rogersteblerbsi replied:
<strike>Is now called in #start before #execLoadSession().

With this change we assume that all client sessions have a backend where they can load the variables from. Should we do that? For now I have overridden the method in the test client sessions that don't require a running backend.</strike>

After discussion with Paolo we keep it as it is because it should be possible to have client sessions without a backend. Client Sessions that want to load data from the backend should explicitly do that by calling `loadAndInitializeVariables`  in `execLoadSession` similar to the way it was done before using `initializeSharedVariables`.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
-  public void replaceSharedVariableMapInternal(Map<String, Object> newMap) {
-    m_sharedVariableMap.updateInternal(newMap);
+  public void replaceVariableMapInternal(Map<String, Object> newMap) {
+    m_variableMap.updateInternal(newMap);
```
Is this method invoked from outside session & subclasses? 
Is this method invoked at all? Without server session there shoudn't be the need of exchanging the whole map any longer. Remove method?

rogersteblerbsi replied:
Removed

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
+  }
+
+  @Override
+  public void replaceVariableInternal(String variableName, Object newValue) {
```
Is this method invoked from outside session & subclasses? If not -> remove from interface and mark protected in AbstractClientSession.

rogersteblerbsi replied:
Removed from interface and changed to `protected`.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
+  }
+
+  @Override
+  public void replaceVariablesInternal(Map<String, Object> variables) {
```
Is this method invoked from outside session & subclasses? If not -> remove from interface and mark protected in AbstractClientSession.

rogersteblerbsi replied:
Removed from interface and changed to `protected`.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -107,8 +109,8 @@ public AbstractClientSession(boolean autoInitConfig) {
     m_objectExtensions = new ObjectExtensions<>(this, true);
-    m_sharedVariableMap = new SharedVariableMap();
-    m_exposedSharedVariables = null;
+    m_variableMap = new SharedVariableMap();
```
Also rename org.eclipse.scout.rt.shared.services.common.context.SharedVariableMap to VariableMap or SessionVariableMap/SessionVariables (as in TS) and move class to client?

rogersteblerbsi replied:
`SharedVariableMap` basically is just a Map with property change listeneners. This could theoretically also be used for other things than the session variables. In our code we just use it for the session. Should we rename and move it anyways?
Edit: Discussed with Paolo: We don't move/rename the `SharedVariableMap`.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
+  }
+
+  /**
+   * Hook for initializing session variables
```
"Hook for initializing custom session variables"

rogersteblerbsi replied:
Changed

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/AbstractServerSession.java:
```diff
@@ -80,27 +72,6 @@ public AbstractServerSession(boolean autoInitConfig) {
-
   private void assignUserId() {
     String userId = BEANS.get(IAccessControlService.class).getUserIdOfCurrentSubject();
     setUserIdInternal(userId);
```
inline private method setUserIdInternal()

rogersteblerbsi replied:
Changed

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -165,7 +167,7 @@ public String getId() {
   @Override
   public String getUserId() {
-    return getSharedContextVariable("userId", String.class);
+    return BEANS.get(IAccessControlService.class).getUserIdOfCurrentSubject();
```
(1)
- Former we used the userId extracted from the subject in backend here. 
- Check with e.g. @andibur if this is guaranteed
- If not guaranteed, add userId to init-call and use backend value here

(2)
- extracting userId from subject is quite expensive and userId is commonly used by session consumers. Cache this value and only resolve once in #start() (or fetch from backend in loadAndInitializeVariables() once providing a sample impl. in scout for this mechanism)

rogersteblerbsi replied:
Loaded now via init-call from backend and set in `loadAndInitializeVariables` as session variable.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,38 @@ protected IMemoryPolicy resolveMemoryPolicy() {
+  /**
+   * Hook for initializing session variables
+   */
+  protected void initializeVariables(LoadInitialVariablesResponse initialVariablesResponse) {
```
according to naming of other methods maybe: `execInitVariables` or `execInitializeVariables` and add to `IClientSessionExtension` API for extensibility? To be discussed....

rogersteblerbsi replied:
Renamed to `execInitVariables` and added to the `IClientSessionExtension`

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/IClientSession.java:
```diff
@@ -134,8 +153,13 @@ public interface IClientSession extends ISession, IPropertyObserver {
+  /**
+   * @return the variable map holding additional information used by the client session
+   */
+  Map<String, Object> getVariableMap();
```
Add release notes section with session API changes.

rogersteblerbsi replied:
Added: https://github.com/eclipse-scout/scout.docs/commit/c14ded9cf8604ea616c3c379226dd18baf24ab98

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/session/ISessionService.java:
```diff
@@ -0,0 +1,19 @@
+import org.eclipse.scout.rt.shared.TunnelToServer;
+
+@TunnelToServer
+public interface ISessionService extends IService {
```
Add javadoc like "Service supporting creation and handling of client sessions"

rogersteblerbsi replied:
JavaDoc added.

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/session/SessionService.java:
```diff
@@ -0,0 +1,38 @@
+  public LoadInitialVariablesResponse loadInitialVariables() {
+    Map<String, Serializable> additionalData = new HashMap<>();
+
+    interceptAdditionalData(additionalData);
```
Consider adding userId here as default value in Scout?

rogersteblerbsi replied:
`userId` now loaded here and set in loadAndInitializeVariables as session variable.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/IClientSession.java:
```diff
@@ -125,7 +125,26 @@ public interface IClientSession extends ISession, IPropertyObserver {
+   * @param newMap
+   *     map to replace the current one with
+   */
+  void replaceVariableMapInternal(Map<String, Object> newMap);
```
see AbstractClientSession, try to remove those three public available replaceXY methods

rogersteblerbsi replied:
Removed

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/ISession.java:
```diff
@@ -33,16 +32,10 @@ public interface ISession {
 
   /**
-   * Shared context variable containing the authenticated userId in lowercase
+   * Authenticated userId in lowercase
```
Suggestion 
```
  /**
   * Authenticated userId, extracted by {@link IAccessControlService#getUserIdOfCurrentSubject()}
   */
```
-> Lowercase is not part of the implementation and therefore not guaranteed.

rogersteblerbsi replied:
Changed

paolobazzi reviewed (commented):
Looks good, see inputs for some minor improvements (or disussion points).

cguglielmo commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -288,26 +289,25 @@ protected IMemoryPolicy resolveMemoryPolicy() {
-  @Override
-  public void replaceSharedVariableMapInternal(Map<String, Object> newMap) {
-    m_sharedVariableMap.updateInternal(newMap);
+  protected void replaceVariableInternal(String variableName, Object newValue) {
```
I think setVariable would be a better name because replace does not sound like it would add new variables. The internal suffix is not necessary as it is already protected. Also add the Java Doc that you already wrote for the interface method. Same applies for the other method.

rogersteblerbsi replied:
Methods renamed and java doc added.

cguglielmo commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/TestServerSession.java:
```diff
@@ -10,9 +10,7 @@
- * A test server session with shared context variables. <br>
- * Has to be located in package <code>org.eclipse.scout.rt.server</code>, because ServerSessionRegistryService requires
- * .* it
+ * A test server session
```
Remove java doc completely as it gives nor more information than the class name itself.

rogersteblerbsi replied:
Removed

cguglielmo commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/services/common/context/SharedVariableMap.java:
Should this be renamed to SessionVariableMap?

rogersteblerbsi replied:
The `SharedVariableMap` is essentially a `Map` with support for property change listeners. Since it's a more general-purpose construct that could be useful beyond just session-related use cases, we decided not to rename the class.

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/services/common/context/SharedVariableMap.java:
```diff
@@ -73,17 +73,18 @@ public synchronized void updateInternal(Map<String, Object> newMap) {
-  private void fireValuesChanged() {
-    m_propertySupport.firePropertyChange(PROP_VALUES, null, CollectionUtility.copyMap(m_variables));
+  private void fireValuesChanged(Map<String, Object> oldVariables) {
+    m_propertySupport.firePropertyChange(PROP_VALUES, oldVariables, CollectionUtility.copyMap(m_variables));
```
Add unit test for this new behavior of including oldValues in property change events.

rogersteblerbsi replied:
added unit test for old variables. additionally added some unit tests for the other methods that fire a change event.

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/session/SessionService.java:
```diff
@@ -0,0 +1,32 @@
+  public LoadInitialSharedVariablesResponse loadInitialSharedVariables() {
+    Map<String, Object> variables = new HashMap<>();
+
+    for (IInitialSharedVariableContributor contributor : BEANS.all(IInitialSharedVariableContributor.class)) {
```
Use lambda-notation:
`BEANS.all(IInitialSharedVariableContributor.class).forEach(c -> c.contribute(variables));`

rogersteblerbsi replied:
updated

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/session/IInitialSharedVariableContributor.java:
```diff
@@ -0,0 +1,20 @@
+import org.eclipse.scout.rt.platform.ApplicationScoped;
+
+@ApplicationScoped
+public interface IInitialSharedVariableContributor {
```
Add javadoc indicating purpose of this interface, e.g. "Contributor for client session variables initially loaded by SessionService"

rogersteblerbsi replied:
javadoc added

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/ISession.java:
```diff
@@ -22,6 +22,8 @@
 @Bean
 public interface ISession {
 
+  String SHARED_CONTEXT_USER_ID = "userId";
```
Since we'll eliminate this interface in near future, should we move this constant to a `ISessionVariable` interface shared between client & server?

rogersteblerbsi replied:
moved to an `ISessionVariable` interface

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/AbstractClientSession.java:
```diff
@@ -289,25 +290,30 @@ protected IMemoryPolicy resolveMemoryPolicy() {
+   * Loads the initial shared variables from the server. Blocks until the initial version of the variables is
    * available or the timeout is reached.
    *
    * @throws ProcessingException
```
remove this throws section in javadoc

rogersteblerbsi replied:
removed

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/context/SharedContextNotificationHandler.java:
```diff
@@ -23,6 +27,11 @@ public class SharedContextNotificationHandler implements INotificationHandler<Sh
-    session.replaceSharedVariableMapInternal(notification.getSharedVariableMap());
+    Map<String, Object> variables = notification.getSharedVariableMap();
+    if (variables == null) {
+      // no changed variables provided in the notification. load the variables from backend.
```
what's the use case for this handler? are there cases, where the server is not able to send all new values to client?

rogersteblerbsi replied:
for example when a complete backend cache (for example a user cache) is invalidated. then we don't want to directly load all data again just to provide it in all the notifications. the still active clients can then ask and fetch the data by themselves.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/clientnotification/ClientSessionRegistry.java:
remove import

rogersteblerbsi replied:
removed

paolobazzi commented on org.eclipse.scout.rt.client.test/src/test/java/org/eclipse/scout/rt/client/ui/basic/table/organizer/OrganizeColumnsFormTest.java:
```diff
@@ -45,19 +38,6 @@
-
-  @Before
-  public void before() {
-    BeanMetaData newBean = new BeanMetaData(AllAccessControlService.class).withOrder(BeanTestingHelper.TESTING_BEAN_ORDER);
```
why is this no longer necessary?

rogersteblerbsi replied:
thanks for the info. was accidentally removed during some refactoring. re-added.

rogersteblerbsi committed:
Avoid stale IdCodec in tests

Ensure that tests which replace the IdCodec use a fresh instance of the
used DataObjectMappers.

This avoids scenarios where AbstractIdCodecSerializer retains the
replaced IdCodec due to internal caching in the DataObjectMapper,
potentially affecting subsequent tests.

423389
